py3status/modules/gitlab.py

```diff
@@ -0,0 +1,163 @@
+
+Format placeholders:
+    {open_merge_requests_count} number of open merge requests, eg 1
+
+    See https://docs.gitlab.com/ee/api/projects.html#get-single-project
+    for explicit list of valid Gitlab placeholders to use. Not all of
+    Gitlab placeholders will be supported. Here's some examples.
+
+    {name}              project name, eg py3status
+    {star_count}        number of stars, eg 2
+    {forks_count}       number of forks, eg 3
+    {open_issues_count} number of open issues, eg 4
+
+    Number of notifications are not available at this time.
+
+Examples:
+```
+# follow a fictional project
+gitlab {
+    auth_token = 'abcdefghijklmnopq-A4'
+    url = 'https://gitlab.com/ultrabug/py3status'
+}
+```
+
+@author lasers
+
+SAMPLE OUTPUT
+{'full_text': 'py3status 48/49'}
+
+"""
+#     format: display format for this module
+#         (default '\?not_zero Gitlab {notification}|{name}
+#         {open_issues_count}/{open_merge_requests_count}')
+
+# notification
+# {'full_text': 'Gitlab 3', 'urgent': True}
+
+# format = ('\?not_zero Gitlab {notification}|{name} '
+#           '{open_issues_count}/{open_merge_requests_count}')
+
+STRING_ERROR_URL = 'missing url'
+STRING_ERROR_AUTH = 'missing auth_token'
+
+try:
+    import urlparse
+except ImportError:
+    import urllib.parse as urlparse
+
+
+class Py3status:
+    """
+    """
+    # available configuration parameters
+    auth_token = None
+    button_action = 1
+    button_refresh = 2
+    cache_timeout = 60
+    format = '{name} {open_issues_count}/{open_merge_requests_count}'
+    url = None
+
+    def post_config_hook(self):
+        if not self.url:
+            raise Exception(STRING_ERROR_URL)
+        if not self.auth_token:
+            raise Exception(STRING_ERROR_AUTH)
+        # url
+        self.button_url = self.url
+        base = self.url.rsplit('/', 2)[0]
+        project = '%2F'.join(self.url.rsplit('/', 2)[1:])
+        self.url = base + '/api/v4/projects/' + project
+        self.open_merge_url = self.url + \
+            '/merge_requests?state=opened&view=simple&per_page=1'
+        # others
+        self.first_run = False
+        self.headers = {'PRIVATE-TOKEN': self.auth_token}
+        self.request_timeout = 10
+
+    def _get_data(self, url):
+        self.count += 1
+        try:
+            return self.py3.request(
+                url, timeout=self.request_timeout, headers=self.headers)
+        except self.py3.RequestException:
+            return {}
+
+    def _get_open_merge_requests_count(self):
+        data = self._get_data(self.open_merge_url)
+        links = data.headers.get('Link')
+        if not links:
+            return len(data.json())
+
+        last_page = 1
+        for link in links.split(','):
+            if 'rel="last"' in link:
+                last_url = link[link.find('<') + 1:link.find('>')]
+                parsed = urlparse.urlparse(last_url)
+                last_page = int(urlparse.parse_qs(parsed.query)['page'][0])
```

Comment:
Can't you just get this from the headers? https://docs.gitlab.com/ce/api/README.html#other-pagination-headers

Reply:
This works. I optimized the f--- out of this module now. All done. I promise... I promise...